File: apps/frontend/Coffee_Call/Coffee_Call/ViewModels/DriftDetailViewModel.swift
import SwiftUI
import Combine
import FirebaseAuth
import FirebaseFirestore

struct ParticipantDetail: Identifiable, Hashable {
    let id = UUID()
    let name: String
    let initials: String
    let interests: [String]
    let joinTimeDescription: String
}

class DriftDetailViewModel: ObservableObject {
    @Published var drift: Drift
    @Published var joinStatus: JoinStatus = .notJoined
    @Published var participants: [ParticipantDetail] = []
    @Published var savedDrifts: [Drift] = []
    @Published var showingRequestSentConfirmation = false
    
    private var cancellables = Set<AnyCancellable>()
    
    enum JoinStatus {
        case notJoined
        case requested
        case joined
        case full
        case ended
    }
    
    var canAccessChat: Bool {
        joinStatus == .joined
    }
    
    private let driftsService: DriftsServiceProtocol
    
    init(drift: Drift, initialJoinStatus: JoinStatus? = nil, driftsService: DriftsServiceProtocol? = nil) {
        if let driftsService = driftsService {
            self.driftsService = driftsService
        } else {
            let isFirebaseEnabled = Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist") != nil
            self.driftsService = isFirebaseEnabled ? FirebaseDriftsService() : MockDriftsService()
        }
        // Build rich mock details for the host to display in the Host Context Card
        let hostName = drift.host.name
        let mockOtherActive = [
            Drift(
                title: "Walk in Indiranagar",
                description: "Morning stroll followed by filter coffee.",
                location: "Indiranagar, Bangalore",
                meetingPoint: "Cult Fit entrance",
                time: "7:00 AM",
                endTime: "8:00 AM",
                date: "Tomorrow",
                distance: 1.5,
                status: .open,
                category: .walk,
                hook: "Morning energy 🚶‍♂️",
                host: Host(name: hostName, role: drift.host.role, imageUrl: drift.host.imageUrl, isVerified: drift.host.isVerified),
                peopleGoing: 2,
                spotsLeft: 3,
                capacity: 5,
                vibeTags: ["Fresh", "Early birds"],
                whatToBring: ["Comfortable footwear"],
                participantInitials: ["SR", "AP"],
                imageUrl: nil
            ),
            Drift(
                title: "Specialty Coffee Tasting",
                description: "Let's explore some local light roasts.",
                location: "Koramangala, Bangalore",
                meetingPoint: "Third Wave Coffee",
                time: "4:00 PM",
                endTime: "5:00 PM",
                date: "Saturday",
                distance: 3.2,
                status: .open,
                category: .coffee,
                hook: "For coffee nerds ☕",
                host: Host(name: hostName, role: drift.host.role, imageUrl: drift.host.imageUrl, isVerified: drift.host.isVerified),
                peopleGoing: 1,
                spotsLeft: 4,
                capacity: 5,
                vibeTags: ["Educational", "Casual"],
                whatToBring: ["An open palate"],
                participantInitials: ["KT"],
                imageUrl: nil
            )
        ]
        
        let richHost = Host(
            id: drift.host.id,
            name: drift.host.name,
            role: drift.host.role,
            imageUrl: drift.host.imageUrl,
            isVerified: drift.host.isVerified,
            hostedCount: 4,
            joinedCount: 12,
            completedCount: 16,
            verified: true,
            otherActiveDrifts: mockOtherActive,
            pastDrifts: ["Walk in Indiranagar", "Coffee chat"],
            interests: ["Walks", "Coffee", "Movies"],
            firestoreUID: drift.host.firestoreUID
        )
        
        var updatedDrift = drift
        updatedDrift.host = richHost
        self.drift = updatedDrift

        // Participants: fetch live from Firestore when Firebase is enabled.
        // Fall back to mock people in offline/preview mode.
        if isFirebaseEnabled {
            self.participants = [] // will be populated by fetchParticipants()
            fetchParticipants()
            fetchHostOtherActiveDrifts()
        } else {
            self.participants = [
                ParticipantDetail(name: "Liam", initials: "LJ", interests: ["Walks", "Coffee", "Music"], joinTimeDescription: "Joined today 2:14 PM"),
                ParticipantDetail(name: "Maya", initials: "MM", interests: ["Coffee", "Walks", "Music"], joinTimeDescription: "Joined today 1:45 PM"),
                ParticipantDetail(name: "Sarah", initials: "SJ", interests: ["Walks", "Coffee", "Music"], joinTimeDescription: "Joined today 12:30 PM"),
                ParticipantDetail(name: "Dev", initials: "DG", interests: ["Music", "Coffee", "Walks"], joinTimeDescription: "Joined today 11:15 AM")
            ]
        }
        
        if let initialJoinStatus {
            self.joinStatus = initialJoinStatus
        } else if drift.status == .ended {
            self.joinStatus = .ended
        } else if drift.isMine {
            self.joinStatus = .joined
        } else if drift.spotsLeft == 0 {
            self.joinStatus = .full
        } else {
            self.joinStatus = .notJoined
        }
        
        setupBookmarkSubscription()
    }
    
    private func setupBookmarkSubscription() {
        BookmarkManager.shared.$savedDrifts
            .receive(on: RunLoop.main)
            .assign(to: \.savedDrifts, on: self)
            .store(in: &cancellables)
    }
    
    func requestToJoin() {
        let currentUid = Auth.auth().currentUser?.uid ?? ""

        // Use real profile data from UserDefaults (written by ProfileViewModel on save/fetch).
        // Falls back to MockData only if UserDefaults has nothing (offline preview mode).
        let currentUserName = UserDefaults.standard.string(forKey: "profile_name")
            ?? AppConstants.MockData.userName
        let currentUserInitials = UserDefaults.standard.string(forKey: "profile_initials")
            ?? AppConstants.MockData.userInitials
        let currentUserRole = "Member"

        let formatter = DateFormatter()
        formatter.timeStyle = .short
        let timestamp = formatter.string(from: Date())

        let joinRequest = JoinRequest(
            userId: currentUid,
            userName: currentUserName,
            userInitials: currentUserInitials,
            userRole: currentUserRole,
            message: "Hey, I'd love to join your drift!",
            timestamp: timestamp
        )

        driftsService.requestToJoin(driftId: drift.id, request: joinRequest)
            .receive(on: RunLoop.main)
            .sink(receiveCompletion: { completionResult in
                if case .failure(let error) = completionResult {
                    print("Error requesting to join: \(error)")
                }
            }, receiveValue: { [weak self] in
                guard let self = self else { return }
                withAnimation(.spring()) {
                    self.joinStatus = .requested
                    self.showingRequestSentConfirmation = true
                }
            })
            .store(in: &cancellables)
    }

    // MARK: - Firebase Participants Fetch
    private var isFirebaseEnabled: Bool {
        Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist") != nil
    }

    func fetchParticipants() {
        guard isFirebaseEnabled else { return }
        let db = Firestore.firestore()
        let postId = drift.id.uuidString

        db.collection("acceptances")
            .whereField("postId", isEqualTo: postId)
            .getDocuments { [weak self] snapshot, error in
                guard let self = self, let docs = snapshot?.documents, error == nil else { return }

                let group = DispatchGroup()
                var fetched: [ParticipantDetail] = []

                for doc in docs {
                    let data = doc.data()
                    guard let acceptorId = data["acceptorId"] as? String else { continue }

                    group.enter()
                    db.collection("users").document(acceptorId).getDocument { userSnap, _ in
                        defer { group.leave() }
                        let userData = userSnap?.data() ?? [:]
                        let name = userData["name"] as? String ?? "Someone"
                        let parts = name.components(separatedBy: " ")
                        let initials = parts.compactMap { $0.first }.map { String($0) }.joined().uppercased()
                        let interests = userData["interestTags"] as? [String] ?? []

                        var joinTimeDesc = "Joined recently"
                        if let ts = data["acceptedAt"] as? Timestamp {
                            let fmt = DateFormatter()
                            fmt.timeStyle = .short
                            joinTimeDesc = "Joined \(fmt.string(from: ts.dateValue()))"
                        }

                        fetched.append(ParticipantDetail(
                            name: name,
                            initials: initials,
                            interests: interests,
                            joinTimeDescription: joinTimeDesc
                        ))
                    }
                }

                group.notify(queue: .main) { [weak self] in
                    self?.participants = fetched
                }
            }
    }
    
    func fetchHostOtherActiveDrifts() {
        guard isFirebaseEnabled else { return }
        let creatorId = drift.host.firestoreUID
        guard !creatorId.isEmpty else { return }
        
        let db = Firestore.firestore()
        db.collection("posts")
            .whereField("creatorId", isEqualTo: creatorId)
            .getDocuments { [weak self] snapshot, error in
                guard let self = self, let docs = snapshot?.documents, error == nil else { return }
                
                let currentUid = Auth.auth().currentUser?.uid
                let fetchedDrifts: [Drift] = docs.compactMap { doc -> Drift? in
                    // Let's skip the current drift!
                    if doc.documentID == self.drift.id.uuidString { return nil }
                    
                    let data = doc.data()
                    let isActive = data["isActive"] as? Bool ?? true
                    if !isActive { return nil }
                    
                    let id = UUID.fromString(doc.documentID)
                    let title = data["title"] as? String ?? ""
                    let description = data["description"] as? String ?? ""
                    let location = data["location"] as? String ?? ""
                    let meetingPoint = data["meetingPoint"] as? String ?? ""
                    let time = data["time"] as? String ?? ""
                    let endTime = data["endTime"] as? String ?? ""
                    let date = data["date"] as? String ?? ""
                    let distance = data["distance"] as? Double ?? 1.2
                    
                    let statusStr = data["status"] as? String ?? "OPEN"
                    let status: DriftStatus
                    switch statusStr.uppercased() {
                    case "OPEN": status = .open
                    case "STARTING SOON": status = .startingSoon
                    case "TONIGHT": status = .tonight
                    case "ENDED": status = .ended
                    default: status = .open
                    }
                    
                    let categoryStr = data["category"] as? String ?? "coffee"
                    let category = DriftCategory(rawValue: categoryStr.lowercased()) ?? .coffee
                    let hook = data["hook"] as? String
                    
                    let hostName = data["creatorName"] as? String ?? "Host"
                    
                    let host = Host(
                        id: UUID.fromString(creatorId),
                        name: hostName,
                        role: "Host",
                        imageUrl: data["creatorImageUrl"] as? String,
                        isVerified: data["creatorVerified"] as? Bool ?? false,
                        firestoreUID: creatorId
                    )
                    
                    let peopleGoing = data["participantCount"] as? Int ?? 1
                    let capacity = data["capacity"] as? Int ?? 5
                    let spotsLeft = data["spotsLeft"] as? Int ?? (capacity - peopleGoing)
                    
                    let vibeTags = data["vibeTags"] as? [String] ?? []
                    let whatToBring = data["whatToBring"] as? [String] ?? []
                    let participantInitials = data["participantInitials"] as? [String] ?? []
                    let imageUrl = data["imageUrl"] as? String
                    
                    let isMine = (creatorId == currentUid)
                    
                    return Drift(
                        id: id,
                        title: title,
                        description: description,
                        location: location,
                        meetingPoint: meetingPoint,
                        time: time,
                        endTime: endTime,
                        date: date,
                        distance: distance,
                        status: status,
                        category: category,
                        hook: hook,
                        host: host,
                        peopleGoing: peopleGoing,
                        spotsLeft: spotsLeft,
                        capacity: capacity,
                        vibeTags: vibeTags,
                        whatToBring: whatToBring,
                        participantInitials: participantInitials,
                        imageUrl: imageUrl,
                        isMine: isMine
                    )
                }
                
                DispatchQueue.main.async {
                    var updatedHost = self.drift.host
                    let newHost = Host(
                        id: updatedHost.id,
                        name: updatedHost.name,
                        role: updatedHost.role,
                        imageUrl: updatedHost.imageUrl,
                        isVerified: updatedHost.isVerified,
                        hostedCount: updatedHost.hostedCount,
                        joinedCount: updatedHost.joinedCount,
                        completedCount: updatedHost.completedCount,
                        verified: updatedHost.verified,
                        otherActiveDrifts: fetchedDrifts,
                        pastDrifts: updatedHost.pastDrifts,
                        interests: updatedHost.interests,
                        firestoreUID: updatedHost.firestoreUID
                    )
                    self.drift.host = newHost
                }
            }
    }
    
    var isBookmarked: Bool {
        savedDrifts.contains(where: { $0.id == drift.id || ($0.title == drift.title && $0.host.name == drift.host.name) })
    }
    
    func saveDrift() {
        BookmarkManager.shared.toggleBookmark(drift)
    }
    
    func shareDrift() {
        let inviteText = AppStrings.Drifts.Detail.inviteText(
            title: drift.title,
            date: drift.date,
            time: drift.time,
            location: drift.location
        )
        UIApplication.shareText(inviteText)
    }
    
    func setReminder() {
        print("Setting reminder...")
    }
}

File: apps/android/app/src/main/java/com/coffeecall/app/feature/driftDetail/DriftDetailViewModel.kt
package com.coffeecall.app.feature.driftDetail

import android.app.Application
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.coffeecall.app.core.firebase.FirebaseUnavailableException
import com.coffeecall.app.core.firebase.FirebaseCollections
import com.coffeecall.app.core.session.SessionPreferencesRepository
import com.coffeecall.app.core.state.GlobalDriftStore
import com.coffeecall.app.data.repository.RepositoryProvider
import com.coffeecall.app.domain.model.DriftPost
import com.coffeecall.app.domain.model.DriftStatus
import com.coffeecall.app.domain.model.JoinMode
import com.coffeecall.app.domain.model.JoinRequest
import com.coffeecall.app.domain.model.UserProfile
import com.coffeecall.app.domain.repository.UserRepository
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.UUID

data class DriftDetailUiState(
    val drift: DriftPost? = null,
    val joinStatus: JoinStatus = JoinStatus.NotJoined,
    val participants: List<ParticipantDetail> = emptyList(),
    val otherActiveDrifts: List<DriftPost> = emptyList(),
    val hostProfile: UserProfile? = null,
    val isLoading: Boolean = false,
    val isActionLoading: Boolean = false,
    val error: String? = null,
    val isReminderSet: Boolean = false,
    val showCalendarConfirmation: Boolean = false,
    val showCalendarDisclaimer: Boolean = false,
    val currentUserId: String = ""
)

enum class JoinStatus {
    NotJoined,
    Requested,
    Joined,
    Full,
    Ended
}

data class ParticipantDetail(
    val name: String,
    val initials: String,
    val interests: List<String>,
    val joinTimeDescription: String = "Joined recently"
)

class DriftDetailViewModel(
    application: Application,
    private val postId: String,
    private val userRepository: UserRepository = RepositoryProvider.userRepository,
    private val sessionRepository: SessionPreferencesRepository = SessionPreferencesRepository(application),
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
    private val store: GlobalDriftStore = GlobalDriftStore.getInstance(application)
) : AndroidViewModel(application) {

    private val _uiState = MutableStateFlow(DriftDetailUiState())
    val uiState: StateFlow<DriftDetailUiState> = _uiState.asStateFlow()

    private val sharedPrefs = application.getSharedPreferences("coffeecall_reminders", Context.MODE_PRIVATE)

    init {
        loadData()
        checkReminderStatus()
        observeStore()
    }

    private fun observeStore() {
        viewModelScope.launch {
            store.drifts.collect { allDrifts ->
                val currentDrift = allDrifts.firstOrNull { it.id == postId } ?: return@collect
                val currentUid = auth.currentUser?.uid ?: ""
                val session = sessionRepository.currentState()
                val userInitials = session.profileInitials.ifBlank { "U" }
                val status = computeJoinStatus(currentDrift, currentUid, userInitials)
                _uiState.update {
                    it.copy(
                        drift = currentDrift,
                        joinStatus = status,
                        currentUserId = currentUid
                    )
                }
            }
        }
    }

    fun loadData() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            val currentUid = auth.currentUser?.uid ?: ""
            _uiState.update { it.copy(currentUserId = currentUid) }

            val session = sessionRepository.currentState()
            val userInitials = session.profileInitials.ifBlank { "U" }

            runCatching {
                val post = store.fetchPost(postId)
                if (post != null) {
                    val status = computeJoinStatus(post, currentUid, userInitials)
                    _uiState.update {
                        it.copy(
                            drift = post,
                            joinStatus = status,
                            isLoading = false
                        )
                    }
                    fetchParticipants(post)
                    fetchHostDetails(post.creatorId)
                } else {
                    loadMockDrift()
                }
            }.onFailure { exception ->
                if (exception is FirebaseUnavailableException) {
                    loadMockDrift()
                } else {
                    _uiState.update { it.copy(isLoading = false, error = exception.localizedMessage ?: "Failed to load post") }
                }
            }
        }
    }

    private fun checkReminderStatus() {
        val key = "calendar_added_$postId"
        val isSet = sharedPrefs.getBoolean(key, false)
        _uiState.update { it.copy(isReminderSet = isSet) }
    }

    fun setReminder() {
        val state = uiState.value
        if (state.isReminderSet) {
            _uiState.update { it.copy(showCalendarDisclaimer = true) }
        } else {
            _uiState.update { it.copy(showCalendarConfirmation = true) }
        }
    }

    fun dismissCalendarConfirmation() {
        _uiState.update { it.copy(showCalendarConfirmation = false) }
    }

    fun dismissCalendarDisclaimer() {
        _uiState.update { it.copy(showCalendarDisclaimer = false) }
    }

    fun confirmAddReminder() {
        _uiState.update { it.copy(showCalendarConfirmation = false, isReminderSet = true) }
        sharedPrefs.edit().putBoolean("calendar_added_$postId", true).apply()
    }

    fun requestToJoin(message: String = "Hey, I'd love to join your drift!") {
        val drift = uiState.value.drift ?: return
        if (uiState.value.joinStatus != JoinStatus.NotJoined) return

        viewModelScope.launch {
            _uiState.update { it.copy(isActionLoading = true) }
            val currentUid = auth.currentUser?.uid ?: ""
            val session = sessionRepository.currentState()
            val currentName = session.profileName.ifBlank { "User" }
            val currentInitials = session.profileInitials.ifBlank { "U" }

            val timestamp = SimpleDateFormat("h:mm a", Locale.getDefault()).format(Date())
            val joinRequest = JoinRequest(
                id = UUID.randomUUID().toString(),
                userId = currentUid,
                userName = currentName,
                userInitials = currentInitials,
                userRole = "Member",
                message = message,
                timestamp = timestamp
            )

            runCatching {
                store.requestToJoin(postId, joinRequest)
                _uiState.update { it.copy(isActionLoading = false) }
            }.onFailure { exception ->
                _uiState.update { it.copy(isActionLoading = false, error = exception.localizedMessage ?: "Action failed") }
            }
        }
    }

    fun cancelJoinRequest() {
        if (uiState.value.joinStatus != JoinStatus.Requested) return

        viewModelScope.launch {
            _uiState.update { it.copy(isActionLoading = true) }
            val currentUid = auth.currentUser?.uid ?: ""

            runCatching {
                store.cancelJoinRequest(postId, currentUid)
                _uiState.update { it.copy(isActionLoading = false) }
            }.onFailure { exception ->
                _uiState.update { it.copy(isActionLoading = false, error = exception.localizedMessage ?: "Failed to cancel request") }
            }
        }
    }

    fun leaveDrift() {
        if (uiState.value.joinStatus != JoinStatus.Joined) return

        viewModelScope.launch {
            _uiState.update { it.copy(isActionLoading = true) }
            val currentUid = auth.currentUser?.uid ?: ""
            val session = sessionRepository.currentState()
            val initials = session.profileInitials.ifBlank { "U" }

            runCatching {
                store.leavePost(postId, currentUid, initials)
                _uiState.update { it.copy(isActionLoading = false) }
            }.onFailure { exception ->
                _uiState.update { it.copy(isActionLoading = false, error = exception.localizedMessage ?: "Failed to leave drift") }
            }
        }
    }

    fun acceptJoinRequest(request: JoinRequest) {
        viewModelScope.launch {
            _uiState.update { it.copy(isActionLoading = true) }
            runCatching {
                store.acceptJoinRequest(postId, request)
                _uiState.update { it.copy(isActionLoading = false) }
            }.onFailure { exception ->
                _uiState.update { it.copy(isActionLoading = false, error = exception.localizedMessage ?: "Failed to accept request") }
            }
        }
    }

    fun rejectJoinRequest(requestId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isActionLoading = true) }
            runCatching {
                store.rejectJoinRequest(postId, requestId)
                _uiState.update { it.copy(isActionLoading = false) }
            }.onFailure { exception ->
                _uiState.update { it.copy(isActionLoading = false, error = exception.localizedMessage ?: "Failed to reject request") }
            }
        }
    }

    fun updateHostDrift(title: String, description: String, location: String) {
        val drift = uiState.value.drift ?: return
        if (drift.creatorId != uiState.value.currentUserId) return

        viewModelScope.launch {
            _uiState.update { it.copy(isActionLoading = true, error = null) }
            runCatching {
                store.upsertPost(
                    drift.copy(
                        title = title.trim().ifBlank { drift.title },
                        description = description.trim(),
                        location = location.trim().ifBlank { drift.location }
                    )
                )
                _uiState.update { it.copy(isActionLoading = false) }
            }.onFailure { exception ->
                _uiState.update { it.copy(isActionLoading = false, error = exception.localizedMessage ?: "Failed to update drift") }
            }
        }
    }

    fun closeDrift() {
        val drift = uiState.value.drift ?: return
        if (drift.creatorId != uiState.value.currentUserId) return

        viewModelScope.launch {
            _uiState.update { it.copy(isActionLoading = true, error = null) }
            runCatching {
                store.updatePostStatus(postId, DriftStatus.Ended)
                _uiState.update { it.copy(isActionLoading = false) }
            }.onFailure { exception ->
                _uiState.update { it.copy(isActionLoading = false, error = exception.localizedMessage ?: "Failed to close drift") }
            }
        }
    }

    fun deleteDrift() {
        val drift = uiState.value.drift ?: return
        if (drift.creatorId != uiState.value.currentUserId) return

        viewModelScope.launch {
            _uiState.update { it.copy(isActionLoading = true, error = null) }
            runCatching {
                val db = FirebaseFirestore.getInstance()
                db.collection(FirebaseCollections.POSTS).document(postId).delete().await()
                db.collection(FirebaseCollections.MESSAGE_THREADS).document(postId).delete().await()
                store.refresh()
                _uiState.update { it.copy(drift = null, isActionLoading = false) }
            }.onFailure { exception ->
                _uiState.update { it.copy(isActionLoading = false, error = exception.localizedMessage ?: "Failed to delete drift") }
            }
        }
    }

    private fun computeJoinStatus(drift: DriftPost, currentUid: String, userInitials: String): JoinStatus {
        val initialsClean = userInitials.trim().uppercase()
        val isMine = drift.creatorId == currentUid

        val isParticipant = drift.participantIds.contains(currentUid) ||
            (drift.participantIds.isEmpty() && drift.participantInitials.any { it.trim().uppercase() == initialsClean })

        val hasPendingRequest = drift.pendingRequests.any { req ->
            if (req.userId.isNotEmpty()) req.userId == currentUid
            else req.userInitials.trim().uppercase() == initialsClean
        }

        return when {
            drift.status == DriftStatus.Ended -> JoinStatus.Ended
            isMine -> JoinStatus.Joined
            isParticipant -> JoinStatus.Joined
            hasPendingRequest -> JoinStatus.Requested
            drift.spotsLeft == 0 -> JoinStatus.Full
            else -> JoinStatus.NotJoined
        }
    }

    private suspend fun fetchParticipants(post: DriftPost) {
        runCatching {
            val db = FirebaseFirestore.getInstance()
            val snap = db.collection("messageThreads").document(postId).get().await()
            if (snap.exists()) {
                val participantsList = snap.get("participants") as? List<*> ?: emptyList<Any>()
                val guestIds = participantsList.mapNotNull { it?.toString() }.filter { it != post.creatorId }

                if (guestIds.isEmpty()) {
                    _uiState.update { it.copy(participants = emptyList()) }
                    return
                }

                val details = mutableListOf<ParticipantDetail>()
                for (uid in guestIds) {
                    val userSnap = db.collection("users").document(uid).get().await()
                    if (userSnap.exists()) {
                        val name = userSnap.getString("name") ?: "Someone"
                        val initials = name.split(" ")
                            .mapNotNull { it.firstOrNull()?.toString() }
                            .joinToString("")
                            .uppercase()
                        val interests = userSnap.get("interestTags") as? List<*> ?: emptyList<Any>()
                        details.add(
                            ParticipantDetail(
                                name = name,
                                initials = if (initials.isEmpty()) "P" else initials,
                                interests = interests.mapNotNull { it?.toString() },
                                joinTimeDescription = "Joined recently"
                            )
                        )
                    }
                }
                _uiState.update { it.copy(participants = details) }
            } else {
                generateFallbackParticipants(post)
            }
        }.onFailure {
            generateFallbackParticipants(post)
        }
    }

    private fun generateFallbackParticipants(post: DriftPost) {
        val namesMap = mapOf(
            "LJ" to "Liam", "MM" to "Maya", "SJ" to "Sarah", "DG" to "Dev",
            "AL" to "Albin", "RI" to "Riya", "MA" to "Maya A.", "SR" to "Sneha", "KM" to "Karthik"
        )
        val interestsMap = mapOf(
            "LJ" to listOf("Walks", "Coffee", "Music"),
            "MM" to listOf("Coffee", "Walks", "Music"),
            "SJ" to listOf("Walks", "Coffee", "Music"),
            "DG" to listOf("Music", "Coffee", "Walks"),
            "AL" to listOf("Walks", "Coffee"),
            "RI" to listOf("Coffee", "Music"),
            "MA" to listOf("Movies", "Music"),
            "SR" to listOf("Walks", "Coffee"),
            "KM" to listOf("Movies", "Coffee")
        )

        val details = post.participantInitials.map { initials ->
            ParticipantDetail(
                name = namesMap[initials] ?: "Member",
                initials = initials,
                interests = interestsMap[initials] ?: listOf("Coffee")
            )
        }
        _uiState.update { it.copy(participants = details) }
    }

    private suspend fun fetchHostDetails(hostId: String) {
        runCatching {
            val profile = userRepository.getUser(hostId)
            if (profile != null) {
                _uiState.update { it.copy(hostProfile = profile) }
            }

            val db = FirebaseFirestore.getInstance()
            val snap = db.collection("posts")
                .whereEqualTo("creatorId", hostId)
                .get()
                .await()

            val otherDrifts = snap.documents.mapNotNull { doc ->
                if (doc.id == postId) return@mapNotNull null
                val data = doc.data ?: return@mapNotNull null
                val statusStr = data["status"] as? String ?: "OPEN"
                if (statusStr.uppercase() == "ENDED") return@mapNotNull null

                val title = data["title"] as? String ?: ""
                val desc = data["description"] as? String ?: ""
                val loc = data["location"] as? String ?: ""
                val time = data["time"] as? String ?: ""
                val date = data["date"] as? String ?: ""

                DriftPost(
                    id = doc.id,
                    title = title,
                    description = desc,
                    location = loc,
                    time = time,
                    date = date,
                    status = DriftStatus.fromFirestore(statusStr)
                )
            }
            _uiState.update { it.copy(otherActiveDrifts = otherDrifts) }
        }
    }

    private fun loadMockDrift() {
        val mockPost = DriftPost(
            id = postId,
            title = "Specialty Coffee Tasting",
            description = "Let's explore some local light roasts and chat about brew methods. Perfect for coffee nerds or beginners!",
            location = "Koramangala, Bangalore",
            meetingPoint = "Third Wave Coffee, Koramangala 4th Block",
            time = "4:00 PM",
            endTime = "5:30 PM",
            date = "Today",
            creatorId = "mock_host_123",
            creatorName = "Siddharth",
            creatorImageUrl = "",
            creatorVerified = true,
            participantCount = 2,
            capacity = 5,
            spotsLeft = 3,
            vibeTags = listOf("Educational", "Casual", "Coffee Nerd"),
            whatToBring = listOf("An open palate", "Questions"),
            participantInitials = listOf("AP", "SR"),
            participantIds = listOf("AP", "SR"),
            latitude = 12.9352,
            longitude = 77.6245,
            joinMode = JoinMode.Approval
        )
        val mockProfile = UserProfile(
            uid = "mock_host_123",
            name = "Siddharth",
            bio = "Always hunting for the cleanest filter coffee. In love with Yirgacheffe profiles.",
            initials = "S",
            interests = listOf("Filter Coffee", "Roasting", "Indie Music"),
            availabilityWeekdayEvenings = true
        )
        val mockOther = listOf(
            DriftPost(
                id = UUID.randomUUID().toString(),
                title = "Morning Stroll & Filter Coffee",
                location = "Indiranagar, Bangalore",
                time = "7:00 AM",
                date = "Tomorrow"
            )
        )

        _uiState.update {
            it.copy(
                drift = mockPost,
                joinStatus = JoinStatus.NotJoined,
                hostProfile = mockProfile,
                otherActiveDrifts = mockOther,
                isLoading = false
            )
        }
        generateFallbackParticipants(mockPost)
    }

    companion object {
        fun factory(application: Application, postId: String): ViewModelProvider.Factory =
            object : ViewModelProvider.Factory {
                @Suppress("UNCHECKED_CAST")
                override fun <T : ViewModel> create(modelClass: Class<T>): T =
                    DriftDetailViewModel(application, postId) as T
            }
    }
}
